tests/pyfunc/test_model_export_with_loader_module_and_data_path.py

```diff
@@ -495,6 +535,99 @@ def test_schema_enforcement_named_tensor_schema_1d():
+        inputs=Schema(
+            [
+                TensorSpec(np.dtype(np.uint64), (-1, 2), "a"),
+                TensorSpec(np.dtype(np.float32), (-1,), "b"),
+            ]
+        )
+    )
+    wrong_pyfunc_model = PyFuncModel(model_meta=wrong_m, model_impl=TestModel())
+    with pytest.raises(
+        expected_exception=MlflowException,
+        match=re.escape(
+            "The input pandas dataframe column 'a' contains scalar "
+            "values, which requires the shape to be (-1,), but got tensor spec "
+            "shape of (-1, 2)."
+        ),
+    ):
+        wrong_pyfunc_model.predict(pdf)
+
+    wrong_m.signature.inputs = Schema(
+        [
+            TensorSpec(np.dtype(np.uint64), (2, -1), "a"),
+            TensorSpec(np.dtype(np.float32), (-1,), "b"),
+        ]
+    )
+    with pytest.raises(
+        expected_exception=MlflowException,
+        match=re.escape(
+            "For pandas dataframe input, the first dimension of shape must be a variable "
+            "dimension and other dimensions must be fixed, but in model signature the shape "
+            "of input a is (2, -1)."
+        ),
+    ):
+        wrong_pyfunc_model.predict(pdf)
+
+    # test that dictionary works too
+    res = pyfunc_model.predict(d_inp)
+    assert res == d_inp
+    expected_types = dict(zip(input_schema.input_names(), input_schema.input_types()))
+    actual_types = {k: v.dtype for k, v in res.items()}
+    assert expected_types == actual_types
+
+
+def test_schema_enforcement_named_tensor_schema_multidimensional():
+    m = Model()
+    input_schema = Schema(
+        [
+            TensorSpec(np.dtype(np.uint64), (-1, 2, 3), "a"),
+            TensorSpec(np.dtype(np.float32), (-1, 3, 4), "b"),
+        ]
+    )
+    m.signature = ModelSignature(inputs=input_schema)
+    pyfunc_model = PyFuncModel(model_meta=m, model_impl=TestModel())
+    data_a = np.array(range(12), dtype=np.uint64)
+    data_b = np.array(range(24), dtype=np.float32) + 10.0
+    pdf = pd.DataFrame(
+        {"a": data_a.reshape(-1, 2 * 3).tolist(), "b": data_b.reshape(-1, 3 * 4).tolist()}
+    )
+    d_inp = {
+        "a": data_a.reshape((-1, 2, 3)),
+        "b": data_b.reshape((-1, 3, 4)),
+    }
+
+    # test dataframe input works for 1d tensor specs and input is converted to dict
+    res = pyfunc_model.predict(pdf)
+    assert _compare_exact_tensor_dict_input(res, d_inp)
+    expected_types = dict(zip(input_schema.input_names(), input_schema.input_types()))
+    actual_types = {k: v.dtype for k, v in res.items()}
+    assert expected_types == actual_types
+
+    wrong_pdf = pdf
+    wrong_pdf.a = np.array(range(16), dtype=np.uint64).reshape(-1, 8).tolist()
+    with pytest.raises(
+        expected_exception=MlflowException,
+        match=re.escape(
+            "The values in Input dataframe column 'a' cannot be converted to expected shape "
+            "(-1, 2, 3) and type uint64"
+        ),
+    ):
+        pyfunc_model.predict(wrong_pdf)
```

Comment:
```suggestion
        pyfunc_model.predict(pdf.assign(a=np.array(range(16), dtype=np.uint64).reshape(-1, 8).tolist()))
```

you can also use `assign`